In [1]:
import pandas as pd
import re

In [2]:

Headers = []
with open('Header.txt', 'r') as p:
    for row in p:
        ye = re.search(r'\s+\b(\D+(\d|))', row)
        header = ye.group(1)
        Headers.append(header.strip())
     

In [3]:
horse = pd.read_csv('horse_data.csv',names=Headers, na_values='?')

In [4]:
horse.describe()

surgery?         Age  Hospital Number  rectal temperature  \
count  299.000000  300.000000     3.000000e+02          240.000000   
mean     1.397993    1.640000     1.085889e+06           38.167917   
std      0.490305    2.173972     1.529801e+06            0.732289   
min      1.000000    1.000000     5.184760e+05           35.400000   
25%      1.000000    1.000000     5.289040e+05           37.800000   
50%      1.000000    1.000000     5.303055e+05           38.200000   
75%      2.000000    1.000000     5.347275e+05           38.500000   
max      2.000000    9.000000     5.305629e+06           40.800000   

            pulse  respiratory rate  temperature of extremities  \
count  276.000000        242.000000                  244.000000   
mean    71.913043         30.417355                    2.348361   
std     28.630557         17.642231                    1.045054   
min     30.000000          8.000000                    1.000000   
25%     48.000000         18.500000                    1.000000   
50%     64.000000         24.500000                    3.000000   
75%     88.000000         36.000000                    3.000000   
max    184.000000         96.000000                    4.000000   

       peripheral pulse  mucous membranes  capillary refill time  ...  \
count        231.000000        253.000000             268.000000  ...   
mean           2.017316          2.853755               1.305970  ...   
std            1.042428          1.620294               0.477629  ...   
min            1.000000          1.000000               1.000000  ...   
25%            1.000000          1.000000               1.000000  ...   
50%            2.000000          3.000000               1.000000  ...   
75%            3.000000          4.000000               2.000000  ...   
max            4.000000          6.000000               3.000000  ...   

       packed cell volume  total protein  abdominocentesis appearance  \
count          271.000000     267.000000                   135.000000   
mean            46.295203      24.456929                     2.037037   
std             10.419335      27.475009                     0.804905   
min             23.000000       3.300000                     1.000000   
25%             38.000000       6.500000                     1.000000   
50%             45.000000       7.500000                     2.000000   
75%             52.000000      57.000000                     3.000000   
max             75.000000      89.000000                     3.000000   

       abdomcentesis total protein     outcome  surgical lesion?  \
count                   102.000000  299.000000        300.000000   
mean                      3.019608    1.551839          1.363333   
std                       1.968567    0.737187          0.481763   
min                       0.100000    1.000000          1.000000   
25%                       2.000000    1.000000          1.000000   
50%                       2.250000    1.000000          1.000000   
75%                       3.900000    2.000000          2.000000   
max                      10.100000    3.000000          2.000000   

       type of lesion1  type of lesion2  type of lesion3     cp_data  
count       300.000000       300.000000       300.000000  300.000000  
mean       3657.880000        90.226667         7.363333    1.670000  
std        5399.513513       649.569234       127.536674    0.470998  
min           0.000000         0.000000         0.000000    1.000000  
25%        2111.750000         0.000000         0.000000    1.000000  
50%        2673.500000         0.000000         0.000000    2.000000  
75%        3209.000000         0.000000         0.000000    2.000000  
max       41110.000000      7111.000000      2209.000000    2.000000  

[8 rows x 28 columns]

In [5]:
"""
Нашли список лошадей, которые посещали клинику несколько раз
"""

horses_count = horse.loc[:,'Hospital Number' ].value_counts().to_frame()
horses_treatment_twice = horses_count.loc[horses_count['Hospital Number'] > 1].reset_index().rename(columns={'index' : 'HospitalNumber'}).HospitalNumber
horses_treatment_twice

0      529796
1      529424
2     5279822
3      527916
4      530526
5      528996
6      528729
7      528151
8      529461
9      528931
10     528890
11     528469
12     530693
13     532349
14     528904
15     527544
Name: HospitalNumber, dtype: int64

Из того что видим, получается что 16 лошадей, повторно привозили в вет.клинику.

In [6]:
#Выберем только необходиммы столбцы
horses = horse.iloc[:,[0,1,3,4,5,10,-6,2]]


## Задание 1

In [7]:
horses.describe()

surgery?         Age  rectal temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       respiratory rate       pain     outcome  Hospital Number  
count        242.000000  245.00000  299.000000     3.000000e+02  
mean          30.417355    2.95102    1.551839     1.085889e+06  
std           17.642231    1.30794    0.737187     1.529801e+06  
min            8.000000    1.00000    1.000000     5.184760e+05  
25%           18.500000    2.00000    1.000000     5.289040e+05  
50%           24.500000    3.00000    1.000000     5.303055e+05  
75%           36.000000    4.00000    2.000000     5.347275e+05  
max           96.000000    5.00000    3.000000     5.305629e+06

In [8]:
horses.mode().head(1)

surgery?  Age  rectal temperature  pulse  respiratory rate  pain  outcome  \
0       1.0  1.0                38.0   48.0              20.0   3.0      1.0   

   Hospital Number  
0           527544

In [9]:

def all_std(horses):
    for column_ in list(horses):
        print(f'{column_} std = {round(horses[column_].std(),2)}')
        print(f'{column_} D = {round(horses[column_].var(),2)}')

In [10]:
all_std(horses)

surgery? std = 0.49
surgery? D = 0.24
Age std = 2.17
Age D = 4.73
rectal temperature std = 0.73
rectal temperature D = 0.54
pulse std = 28.63
pulse D = 819.71
respiratory rate std = 17.64
respiratory rate D = 311.25
pain std = 1.31
pain D = 1.71
outcome std = 0.74
outcome D = 0.54
Hospital Number std = 1529800.91
Hospital Number D = 2340290821346.61


Из того что видно, обычно пациентами оказывались преимущественно взрослые лошади (мода и медиана age = 1, среднее из-за выбросов можно не учитывать), которым делались(пришлось делать?) какие то операции(surgery так же =1), поступившие с прерывистыми слабыми болями  (pain=2  (intermittent mild pain)) и нормальной температурой (Нормальная температура у лошади 37,5-38,5) и у которых в основном был благоприятных исход (outcome = 1). Можно отметить достаточно большие значения ско у пульса и частоты дыхания.

In [11]:
#Найдем дубли
horses_doubles = horses.loc[horses['Hospital Number'].isin(horses_treatment_twice)].sort_values(by='Hospital Number')
horses_doubles

surgery?  Age  rectal temperature  pulse  respiratory rate  pain  \
282       1.0    1                 NaN   70.0              16.0   2.0   
110       1.0    1                 NaN   70.0              16.0   2.0   
166       1.0    1                38.3   40.0              18.0   3.0   
160       1.0    1                 NaN    NaN               NaN   4.0   
179       2.0    1                38.5  100.0               NaN   4.0   
205       2.0    1                38.5  100.0               NaN   4.0   
235       1.0    1                38.1   52.0              24.0   4.0   
73        1.0    1                38.1   52.0              24.0   4.0   
94        2.0    1                37.5   48.0              30.0   NaN   
234       2.0    1                37.5   48.0              30.0   NaN   
291       1.0    1                38.9   80.0              44.0   2.0   
48        1.0    1                38.9   80.0              44.0   2.0   
59        2.0    1                 NaN   96.0               NaN   5.0   
245       2.0    1                 NaN   96.0               NaN   5.0   
67        2.0    1                38.4   54.0              24.0   1.0   
225       2.0    1                38.4   54.0              24.0   1.0   
156       2.0    1                38.6   52.0              20.0   3.0   
38        2.0    1                38.6   52.0               NaN   3.0   
71        1.0    1                37.8   48.0              28.0   1.0   
143       1.0    1                37.8   48.0              28.0   1.0   
91        2.0    1                40.3  114.0              36.0   2.0   
54        2.0    1                40.3  114.0              36.0   2.0   
272       1.0    1                 NaN  100.0              30.0   5.0   
34        1.0    1                 NaN  100.0              30.0   5.0   
233       1.0    1                38.3   72.0              30.0   3.0   
26        1.0    1                38.3   72.0              30.0   3.0   
78        1.0    1                 NaN    NaN               NaN   5.0   
136       1.0    1                 NaN    NaN               NaN   5.0   
138       2.0    1                38.0   56.0               NaN   1.0   
114       1.0    1                38.4   44.0              24.0   5.0   
83        1.0    1                38.0    NaN              24.0   5.0   
182       1.0    1                38.0    NaN              24.0   5.0   

     outcome  Hospital Number  
282      2.0           527544  
110      2.0           527544  
166      1.0           527916  
160      1.0           527916  
179      3.0           528151  
205      3.0           528151  
235      2.0           528469  
73       2.0           528469  
94       1.0           528729  
234      1.0           528729  
291      2.0           528890  
48       2.0           528890  
59       2.0           528904  
245      2.0           528904  
67       1.0           528931  
225      1.0           528931  
156      1.0           528996  
38       1.0           528996  
71       1.0           529424  
143      1.0           529424  
91       2.0           529461  
54       3.0           529461  
272      1.0           529796  
34       1.0           529796  
233      1.0           530526  
26       1.0           530526  
78       2.0           530693  
136      2.0           530693  
138      1.0           532349  
114      1.0           532349  
83       2.0          5279822  
182      2.0          5279822

Проанализировав статистику лошадей, которые дважды попали в больницу, видно что во всех случаях данные скорее всего задвоены, либо в данные отсутствуют. Это позволяет предположить  так же повторные случаи. Кроме одного случая 532349, но в рамках общего кол-ва можно пренебречь.

In [12]:
horses = horses.drop_duplicates(subset = 'Hospital Number',  keep='first')

In [13]:
horses

surgery?  Age  rectal temperature  pulse  respiratory rate  pain  \
0         2.0    1                38.5   66.0              28.0   5.0   
1         1.0    1                39.2   88.0              20.0   3.0   
2         2.0    1                38.3   40.0              24.0   3.0   
3         1.0    9                39.1  164.0              84.0   2.0   
4         2.0    1                37.3  104.0              35.0   NaN   
..        ...  ...                 ...    ...               ...   ...   
295       1.0    1                 NaN  120.0              70.0   2.0   
296       2.0    1                37.2   72.0              24.0   4.0   
297       1.0    1                37.5   72.0              30.0   4.0   
298       1.0    1                36.5  100.0              24.0   3.0   
299       1.0    1                37.2   40.0              20.0   NaN   

     outcome  Hospital Number  
0        2.0           530101  
1        3.0           534817  
2        1.0           530334  
3        2.0          5290409  
4        2.0           530255  
..       ...              ...  
295      3.0           533886  
296      3.0           527702  
297      2.0           529386  
298      1.0           530612  
299      3.0           534618  

[284 rows x 8 columns]

Получили датафрейм без дублей

In [14]:
# Отфильтруем только по плохим исходам
horses[(horses.outcome == 2) | (horses.outcome == 3)].describe()

surgery?         Age  rectal temperature       pulse  \
count  113.000000  113.000000           82.000000  103.000000   
mean     1.292035    1.849558           38.091463   85.048544   
std      0.456724    2.475672            0.951957   30.048451   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.500000   64.000000   
50%      1.000000    1.000000           38.000000   84.000000   
75%      2.000000    1.000000           38.675000  100.000000   
max      2.000000    9.000000           40.800000  184.000000   

       respiratory rate       pain     outcome  Hospital Number  
count         89.000000  92.000000  113.000000     1.130000e+02  
mean          32.505618   3.554348    2.380531     1.204271e+06  
std           16.845638   1.243569    0.487680     1.666470e+06  
min           12.000000   1.000000    2.000000     5.184760e+05  
25%           20.000000   2.000000    2.000000     5.288900e+05  
50%           28.000000   4.000000    2.000000     5.302390e+05  
75%           40.000000   5.000000    3.000000     5.346240e+05  
max           96.000000   5.000000    3.000000     5.301219e+06

In [15]:
horses[(horses.outcome == 2) | (horses.outcome == 3)].mode().head(1)

surgery?  Age  rectal temperature  pulse  respiratory rate  pain  outcome  \
0       1.0  1.0                38.0   88.0              20.0   5.0      2.0   

   Hospital Number  
0           518476

In [16]:
#И только выздоровевших
horses[(horses.outcome == 1)].describe()

surgery?         Age  rectal temperature       pulse  \
count  170.000000  170.000000          145.000000  158.000000   
mean     1.464706    1.564706           38.193793   63.658228   
std      0.500226    2.055140            0.577617   24.815952   
min      1.000000    1.000000           36.500000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   54.000000   
75%      2.000000    1.000000           38.500000   72.000000   
max      2.000000    9.000000           39.900000  150.000000   

       respiratory rate        pain  outcome  Hospital Number  
count        140.000000  137.000000    170.0     1.700000e+02  
mean          29.457143    2.525547      1.0     1.034887e+06  
std           18.698044    1.163667      0.0     1.467886e+06  
min            8.000000    1.000000      1.0     5.216810e+05  
25%           16.000000    2.000000      1.0     5.290652e+05  
50%           24.000000    3.000000      1.0     5.304585e+05  
75%           36.000000    3.000000      1.0     5.347870e+05  
max           96.000000    5.000000      1.0     5.305629e+06

In [17]:
horses[(horses.outcome == 1)].mode().head(1)

surgery?  Age  rectal temperature  pulse  respiratory rate  pain  outcome  \
0       1.0  1.0                38.0   48.0              12.0   3.0      1.0   

   Hospital Number  
0           521681

Сравним показатели выздоровевших (outcome = 1) c результатам плохого исхода (outcome = 2,3). Как видно из данно статистики(113\284) почти 40% животных погибло. Из данных можно увидеть что у погибших выделялись следующие показатели: повышеные болевые ощущения(выше на 1-2 пункта относительно выздоровевших + немного больший разброс относительно среднего ско = 29,6 против 24), значительно повышенным пульсом(причем и мода и медиана и средние значения, относительно выздоровевших мода 88, медиана 84 и среднее 85 против 50-60 у выздоровевших),а так же немного повышенной частотой дыхания, при этом показатели температуры были +- в норме.

In [18]:
# В описании сказано что возможно два значения у атрибута возраст 1 и 2, но 2 не указано, а указано 9.
#Предположим что 9 это молодые, а 1 возрослые.
horses[(horses.Age == 9)].describe()


surgery?   Age  rectal temperature       pulse  respiratory rate  \
count  24.000000  24.0           20.000000   22.000000         20.000000   
mean    1.250000   9.0           38.645000  123.363636         54.600000   
std     0.442326   0.0            0.490408   28.628234         23.904195   
min     1.000000   9.0           38.000000   84.000000         20.000000   
25%     1.000000   9.0           38.275000  100.000000         34.500000   
50%     1.000000   9.0           38.600000  124.000000         49.000000   
75%     1.250000   9.0           38.925000  144.500000         74.000000   
max     2.000000   9.0           39.700000  184.000000         96.000000   

            pain    outcome  Hospital Number  
count  17.000000  24.000000     2.400000e+01  
mean    3.117647   1.541667     4.698006e+06  
std     1.218726   0.588230     1.607492e+06  
min     1.000000   1.000000     5.340920e+05  
25%     2.000000   1.000000     5.287254e+06  
50%     3.000000   1.500000     5.291369e+06  
75%     4.000000   2.000000     5.295194e+06  
max     5.000000   3.000000     5.305629e+06

In [19]:
horses[(horses.Age == 9)].mode().head(1)



surgery?  Age  rectal temperature  pulse  respiratory rate  pain  outcome  \
0       1.0  9.0                38.8  100.0              28.0   2.0      1.0   

   Hospital Number  
0           534092

In [20]:
horses[(horses.Age == 1)].describe()

surgery?    Age  rectal temperature       pulse  respiratory rate  \
count  259.000000  260.0          208.000000  240.000000        210.000000   
mean     1.409266    1.0           38.109135   67.300000         28.309524   
std      0.492650    0.0            0.736268   23.971427         15.576464   
min      1.000000    1.0           35.400000   30.000000          8.000000   
25%      1.000000    1.0           37.700000   48.000000         18.000000   
50%      1.000000    1.0           38.100000   60.000000         24.000000   
75%      2.000000    1.0           38.500000   84.000000         36.000000   
max      2.000000    1.0           40.800000  140.000000         96.000000   

             pain     outcome  Hospital Number  
count  213.000000  259.000000     2.600000e+02  
mean     2.929577    1.552124     7.684457e+05  
std      1.303015    0.757511     1.037408e+06  
min      1.000000    1.000000     5.184760e+05  
25%      2.000000    1.000000     5.288100e+05  
50%      3.000000    1.000000     5.302405e+05  
75%      4.000000    2.000000     5.343438e+05  
max      5.000000    3.000000     5.299603e+06

In [21]:
horses[(horses.Age == 1)].mode().head(1)

surgery?  Age  rectal temperature  pulse  respiratory rate  pain  outcome  \
0       1.0  1.0                38.0   48.0              20.0   3.0      1.0   

   Hospital Number  
0           518476

Видим интересную особенность, что у молодых особоей показатели пульса почти в 2 раза(на 50-60 пунктов) выше чем у взрослых особей и разброс относительно среднего так же выше ~24 против 29, а так же сильно повышенную частоту дыхания (30-50 против 20-30). Так же стоит отметить повышенную относильно верхней границы норму температуру у молодых особей (выше 38,5) при более низком показале ско 0,49 против 0,73.

In [22]:
horses[(horses.Age == 9) ].outcome.value_counts().to_frame() 


outcome
1.0       12
2.0       11
3.0        1

In [23]:
horses[(horses.Age == 1) ].outcome.value_counts().to_frame() 

outcome
1.0      158
2.0       59
3.0       42

Так же становится ясно, что половина молодых особей погибла. У взрослых этот показатель выжиаваемости достигает ~65%

## Задание 2

Посчитаем выбросы по основным аттрибутам

In [24]:
horses.describe()

surgery?         Age  rectal temperature       pulse  \
count  283.00000  284.000000          228.000000  262.000000   
mean     1.39576    1.676056           38.156140   72.007634   
std      0.48988    2.229102            0.733171   28.893382   
min      1.00000    1.000000           35.400000   30.000000   
25%      1.00000    1.000000           37.800000   48.000000   
50%      1.00000    1.000000           38.200000   64.000000   
75%      2.00000    1.000000           38.500000   88.000000   
max      2.00000    9.000000           40.800000  184.000000   

       respiratory rate        pain     outcome  Hospital Number  
count        230.000000  230.000000  283.000000     2.840000e+02  
mean          30.595652    2.943478    1.551237     1.100521e+06  
std           17.998956    1.295382    0.743785     1.547196e+06  
min            8.000000    1.000000    1.000000     5.184760e+05  
25%           18.000000    2.000000    1.000000     5.289738e+05  
50%           24.500000    3.000000    1.000000     5.303570e+05  
75%           36.000000    4.000000    2.000000     5.347848e+05  
max           96.000000    5.000000    3.000000     5.305629e+06

In [25]:
# Найдем выбросы через межквартильный размах
def outliers(column):
    q1 = horses[column].quantile(0.25)
    q3 = horses[column].quantile(0.75)
    iqr = q3 - q1
    lower_bound = q1 - (1.5 * iqr) 
    upper_bound = q3 + (1.5 * iqr)
    remove_outliers = horses[horses[column].between(lower_bound, upper_bound, inclusive=True)]
    return remove_outliers

In [26]:
outliers('rectal temperature')

surgery?  Age  rectal temperature  pulse  respiratory rate  pain  \
0         2.0    1                38.5   66.0              28.0   5.0   
1         1.0    1                39.2   88.0              20.0   3.0   
2         2.0    1                38.3   40.0              24.0   3.0   
3         1.0    9                39.1  164.0              84.0   2.0   
4         2.0    1                37.3  104.0              35.0   NaN   
..        ...  ...                 ...    ...               ...   ...   
292       1.0    1                37.0   66.0              20.0   4.0   
294       2.0    1                38.5   40.0              16.0   2.0   
296       2.0    1                37.2   72.0              24.0   4.0   
297       1.0    1                37.5   72.0              30.0   4.0   
299       1.0    1                37.2   40.0              20.0   NaN   

     outcome  Hospital Number  
0        2.0           530101  
1        3.0           534817  
2        1.0           530334  
3        2.0          5290409  
4        2.0           530255  
..       ...              ...  
292      2.0           530034  
294      1.0           533902  
296      3.0           527702  
297      2.0           529386  
299      3.0           534618  

[215 rows x 8 columns]

In [27]:
outliers('rectal temperature')

surgery?  Age  rectal temperature  pulse  respiratory rate  pain  \
0         2.0    1                38.5   66.0              28.0   5.0   
1         1.0    1                39.2   88.0              20.0   3.0   
2         2.0    1                38.3   40.0              24.0   3.0   
3         1.0    9                39.1  164.0              84.0   2.0   
4         2.0    1                37.3  104.0              35.0   NaN   
..        ...  ...                 ...    ...               ...   ...   
292       1.0    1                37.0   66.0              20.0   4.0   
294       2.0    1                38.5   40.0              16.0   2.0   
296       2.0    1                37.2   72.0              24.0   4.0   
297       1.0    1                37.5   72.0              30.0   4.0   
299       1.0    1                37.2   40.0              20.0   NaN   

     outcome  Hospital Number  
0        2.0           530101  
1        3.0           534817  
2        1.0           530334  
3        2.0          5290409  
4        2.0           530255  
..       ...              ...  
292      2.0           530034  
294      1.0           533902  
296      3.0           527702  
297      2.0           529386  
299      3.0           534618  

[215 rows x 8 columns]

In [28]:
horses.describe()

surgery?         Age  rectal temperature       pulse  \
count  283.00000  284.000000          228.000000  262.000000   
mean     1.39576    1.676056           38.156140   72.007634   
std      0.48988    2.229102            0.733171   28.893382   
min      1.00000    1.000000           35.400000   30.000000   
25%      1.00000    1.000000           37.800000   48.000000   
50%      1.00000    1.000000           38.200000   64.000000   
75%      2.00000    1.000000           38.500000   88.000000   
max      2.00000    9.000000           40.800000  184.000000   

       respiratory rate        pain     outcome  Hospital Number  
count        230.000000  230.000000  283.000000     2.840000e+02  
mean          30.595652    2.943478    1.551237     1.100521e+06  
std           17.998956    1.295382    0.743785     1.547196e+06  
min            8.000000    1.000000    1.000000     5.184760e+05  
25%           18.000000    2.000000    1.000000     5.289738e+05  
50%           24.500000    3.000000    1.000000     5.303570e+05  
75%           36.000000    4.000000    2.000000     5.347848e+05  
max           96.000000    5.000000    3.000000     5.305629e+06

In [29]:
outliers('respiratory rate')

surgery?  Age  rectal temperature  pulse  respiratory rate  pain  \
0         2.0    1                38.5   66.0              28.0   5.0   
1         1.0    1                39.2   88.0              20.0   3.0   
2         2.0    1                38.3   40.0              24.0   3.0   
4         2.0    1                37.3  104.0              35.0   NaN   
6         1.0    1                37.9   48.0              16.0   3.0   
..        ...  ...                 ...    ...               ...   ...   
294       2.0    1                38.5   40.0              16.0   2.0   
296       2.0    1                37.2   72.0              24.0   4.0   
297       1.0    1                37.5   72.0              30.0   4.0   
298       1.0    1                36.5  100.0              24.0   3.0   
299       1.0    1                37.2   40.0              20.0   NaN   

     outcome  Hospital Number  
0        2.0           530101  
1        3.0           534817  
2        1.0           530334  
4        2.0           530255  
6        1.0           526802  
..       ...              ...  
294      1.0           533902  
296      3.0           527702  
297      2.0           529386  
298      1.0           530612  
299      3.0           534618  

[213 rows x 8 columns]

In [30]:
outliers('pulse')

surgery?  Age  rectal temperature  pulse  respiratory rate  pain  \
0         2.0    1                38.5   66.0              28.0   5.0   
1         1.0    1                39.2   88.0              20.0   3.0   
2         2.0    1                38.3   40.0              24.0   3.0   
4         2.0    1                37.3  104.0              35.0   NaN   
6         1.0    1                37.9   48.0              16.0   3.0   
..        ...  ...                 ...    ...               ...   ...   
295       1.0    1                 NaN  120.0              70.0   2.0   
296       2.0    1                37.2   72.0              24.0   4.0   
297       1.0    1                37.5   72.0              30.0   4.0   
298       1.0    1                36.5  100.0              24.0   3.0   
299       1.0    1                37.2   40.0              20.0   NaN   

     outcome  Hospital Number  
0        2.0           530101  
1        3.0           534817  
2        1.0           530334  
4        2.0           530255  
6        1.0           526802  
..       ...              ...  
295      3.0           533886  
296      3.0           527702  
297      2.0           529386  
298      1.0           530612  
299      3.0           534618  

[257 rows x 8 columns]

Как видно что довольно много выбросов у показателей частота дыхания и температура. По имеющимся данным мы не можем четко сказать что выбросы были ошибочный, т.к. например пульс лошади может зависеть от породы и ее предназначения(во время забегов у некоторых может быть больше 250), так же и частота дыхания и температура. Поэтому нет причин их не учитывать.

## Задание 3

In [31]:
for col in horses.columns:
    pct_missing = horses[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery? - 0.4%
Age - 0.0%
rectal temperature - 19.7%
pulse - 7.7%
respiratory rate - 19.0%
pain - 19.0%
outcome - 0.4%
Hospital Number - 0.0%


In [32]:
horses.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 284 entries, 0 to 299
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   surgery?            283 non-null    float64
 1   Age                 284 non-null    int64  
 2   rectal temperature  228 non-null    float64
 3   pulse               262 non-null    float64
 4   respiratory rate    230 non-null    float64
 5   pain                230 non-null    float64
 6   outcome             283 non-null    float64
 7   Hospital Number     284 non-null    int64  
dtypes: float64(6), int64(2)
memory usage: 20.0 KB


Как видно выше, в столбцах 'rectal temperature', 'respiratory rate' и 'pain' достаточно много пропусков. Но так как и мода и медиана и среднее по температуре находятся довольно близко, то я решил заменить недостающие значения с помощью среднего

In [33]:
horses_na = pd.read_csv('horse_data.csv',names=Headers, na_values='?')
horses_na_eight = horses_na.iloc[:,[0,1,3,4,5,10,-6,2]]
horses_na_no_dubl = horses_na_eight.drop_duplicates(subset = 'Hospital Number',  keep='first')


In [34]:
"""
Так как значение показателей моды, медианы и среднего были довольно близки, то заполнил пропуски значением среднего.

"""
horses_na_no_dubl['rectal temperature'].describe()


count    228.000000
mean      38.156140
std        0.733171
min       35.400000
25%       37.800000
50%       38.200000
75%       38.500000
max       40.800000
Name: rectal temperature, dtype: float64

In [35]:
horses_na_no_dubl['rectal temperature'].mode()[0]

38.0

In [36]:
horses_na_no_dubl.iloc[:, 2].fillna(horses['rectal temperature'].mean(), inplace = True)

In [37]:
"""
С пульсом оказалось все сложнее, был очень большой разброс,
поэтому пришлось группировать по Age и outcome и заполнять медианными значениями.
"""
horses_na_no_dubl.pulse.describe()

count    262.000000
mean      72.007634
std       28.893382
min       30.000000
25%       48.000000
50%       64.000000
75%       88.000000
max      184.000000
Name: pulse, dtype: float64

In [38]:
horses_na_no_dubl.pulse.mode()[0]

48.0

In [39]:
horses_na_no_dubl.iloc[:, 3].fillna(horses.groupby(['Age','outcome']).pulse.transform('median'), inplace=True)

In [40]:
"""
C уровнем боли оказлось все проще, и мода медиана и среднее оказлись ~ 3
"""
horses_na_no_dubl.pain.describe()


count    230.000000
mean       2.943478
std        1.295382
min        1.000000
25%        2.000000
50%        3.000000
75%        4.000000
max        5.000000
Name: pain, dtype: float64

In [41]:
horses_na_no_dubl.pain.mode()[0]

3.0

In [42]:
horses_na_no_dubl.iloc[:, 5].fillna(horses.pain.mode()[0], inplace = True)

In [43]:
"""
Частоту дыхания заполняем так же в разбивки по возрасту и исходу
"""

'\nЧастоту дыхания заполняем так же в разбивки по возрасту и исходу\n'

In [44]:
horses_na_no_dubl['respiratory rate'].describe()

count    230.000000
mean      30.595652
std       17.998956
min        8.000000
25%       18.000000
50%       24.500000
75%       36.000000
max       96.000000
Name: respiratory rate, dtype: float64

In [45]:
horses_na_no_dubl['respiratory rate'].mode()[0]

20.0

In [53]:
horses_na_no_dubl.iloc[:, 4].fillna(horses.groupby(['Age','outcome'])['respiratory rate'].transform('median'), inplace=True)

In [47]:
horses_na_no_dubl = horses_na_no_dubl.dropna(subset=['surgery?'])

In [54]:
horses_na_no_dubl.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 283 entries, 0 to 299
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   surgery?            283 non-null    float64
 1   Age                 283 non-null    int64  
 2   rectal temperature  283 non-null    float64
 3   pulse               283 non-null    float64
 4   respiratory rate    282 non-null    float64
 5   pain                283 non-null    float64
 6   outcome             283 non-null    float64
 7   Hospital Number     283 non-null    int64  
dtypes: float64(6), int64(2)
memory usage: 19.9 KB


In [61]:
"""
Почему то осталось одно значение в respiratory rate, которое осталось пустым, возможно изза группировки
"""
horses_na_no_dubl.iloc[75,]

surgery?                   1.0
Age                        9.0
rectal temperature        39.7
pulse                    100.0
respiratory rate           NaN
pain                       2.0
outcome                    3.0
Hospital Number       534092.0
Name: 75, dtype: float64

In [63]:
#Заменим на значение медианы
horses_na_no_dubl.iloc[75,4] = 24.5

In [65]:
for col in horses_na_no_dubl.columns:
    pct_missing = horses_na_no_dubl[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery? - 0.0%
Age - 0.0%
rectal temperature - 0.0%
pulse - 0.0%
respiratory rate - 0.0%
pain - 0.0%
outcome - 0.0%
Hospital Number - 0.0%


Итоговые сравнения двух дата сетов:

In [66]:
horses.describe()

surgery?         Age  rectal temperature       pulse  \
count  283.00000  284.000000          228.000000  262.000000   
mean     1.39576    1.676056           38.156140   72.007634   
std      0.48988    2.229102            0.733171   28.893382   
min      1.00000    1.000000           35.400000   30.000000   
25%      1.00000    1.000000           37.800000   48.000000   
50%      1.00000    1.000000           38.200000   64.000000   
75%      2.00000    1.000000           38.500000   88.000000   
max      2.00000    9.000000           40.800000  184.000000   

       respiratory rate        pain     outcome  Hospital Number  
count        230.000000  230.000000  283.000000     2.840000e+02  
mean          30.595652    2.943478    1.551237     1.100521e+06  
std           17.998956    1.295382    0.743785     1.547196e+06  
min            8.000000    1.000000    1.000000     5.184760e+05  
25%           18.000000    2.000000    1.000000     5.289738e+05  
50%           24.500000    3.000000    1.000000     5.303570e+05  
75%           36.000000    4.000000    2.000000     5.347848e+05  
max           96.000000    5.000000    3.000000     5.305629e+06

In [67]:
horses_na_no_dubl.describe()

surgery?         Age  rectal temperature       pulse  \
count  283.00000  283.000000          283.000000  283.000000   
mean     1.39576    1.678445           38.156692   72.095406   
std      0.48988    2.232686            0.657733   28.558291   
min      1.00000    1.000000           35.400000   30.000000   
25%      1.00000    1.000000           37.850000   48.000000   
50%      1.00000    1.000000           38.156140   64.000000   
75%      2.00000    1.000000           38.500000   88.000000   
max      2.00000    9.000000           40.800000  184.000000   

       respiratory rate        pain     outcome  Hospital Number  
count        283.000000  283.000000  283.000000     2.830000e+02  
mean          29.955830    2.950530    1.551237     1.102521e+06  
std           16.630603    1.165867    0.743785     1.549569e+06  
min            8.000000    1.000000    1.000000     5.184760e+05  
25%           20.000000    2.000000    1.000000     5.289705e+05  
50%           24.000000    3.000000    1.000000     5.303540e+05  
75%           36.000000    4.000000    2.000000     5.347855e+05  
max           96.000000    5.000000    3.000000     5.305629e+06

In [68]:
#Итоговая разница. (Прикольно не знал что можно так вычитать:)
horses_na_no_dubl.describe() - horses.describe()

surgery?       Age  rectal temperature      pulse  respiratory rate  \
count       0.0 -1.000000           55.000000  21.000000         53.000000   
mean        0.0  0.002389            0.000552   0.087773         -0.639822   
std         0.0  0.003585           -0.075437  -0.335091         -1.368354   
min         0.0  0.000000            0.000000   0.000000          0.000000   
25%         0.0  0.000000            0.050000   0.000000          2.000000   
50%         0.0  0.000000           -0.043860   0.000000         -0.500000   
75%         0.0  0.000000            0.000000   0.000000          0.000000   
max         0.0  0.000000            0.000000   0.000000          0.000000   

            pain  outcome  Hospital Number  
count  53.000000      0.0        -1.000000  
mean    0.007052      0.0      1999.820846  
std    -0.129515      0.0      2373.082938  
min     0.000000      0.0         0.000000  
25%     0.000000      0.0        -3.250000  
50%     0.000000      0.0        -3.000000  
75%     0.000000      0.0         0.750000  
max     0.000000      0.0         0.000000

In [70]:
horses_na_no_dubl.mode().head(1) - horses.mode().head(1)

surgery?  Age  rectal temperature  pulse  respiratory rate  pain  outcome  \
0       0.0  0.0             0.15614    0.0               4.0   0.0      0.0   

   Hospital Number  
0                0

По моему мнению, замены были произведены нормально, кроме respiratory rate. Возможно нужно было брать либо другую разбивку, либо не медиану.